File: frontend/src/api/budgets.ts
import { api } from "./client"
import type { BudgetPeriod, BudgetResponse } from "./types"

export const budgetsApi = {
  list: (params: { category_id?: string; effective_date?: string } = {}) => {
    const qs = new URLSearchParams()
    if (params.category_id) qs.set("category_id", params.category_id)
    if (params.effective_date) qs.set("effective_date", params.effective_date)
    const suffix = qs.toString() ? `?${qs.toString()}` : ""
    return api.get<BudgetResponse[]>(`/budgets${suffix}`)
  },

  create: (data: {
    category_id: string
    period?: BudgetPeriod
    amount: string
    effective_from: string
    effective_to?: string | null
  }) => api.post<BudgetResponse>("/budgets", data),

  update: (
    id: string,
    data: Partial<{ amount: string; effective_from: string; effective_to: string | null }>,
  ) => api.patch<BudgetResponse>(`/budgets/${id}`, data),

  delete: (id: string) => api.delete(`/budgets/${id}`),
}

File: backend/app/api/v1/budgets.py
import uuid
from datetime import date

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.visibility import VisibilityContext, get_visibility_ctx
from app.db.base import get_session
from app.schemas.budget import BudgetCreate, BudgetResponse, BudgetUpdate
from app.services.budget import BudgetService

router = APIRouter()


@router.get("/budgets", response_model=list[BudgetResponse])
async def list_budgets(
    category_id: uuid.UUID | None = Query(None),
    effective_date: date | None = Query(None),
    ctx: VisibilityContext = Depends(get_visibility_ctx),
    session: AsyncSession = Depends(get_session),
) -> list[BudgetResponse]:
    svc = BudgetService(session)
    budgets = await svc.list_budgets(ctx, category_id=category_id, effective_date=effective_date)
    return [BudgetResponse.model_validate(b) for b in budgets]


@router.post("/budgets", response_model=BudgetResponse, status_code=201)
async def create_budget(
    data: BudgetCreate,
    ctx: VisibilityContext = Depends(get_visibility_ctx),
    session: AsyncSession = Depends(get_session),
) -> BudgetResponse:
    svc = BudgetService(session)
    budget = await svc.create(ctx, data)
    await session.commit()
    await session.refresh(budget)
    return BudgetResponse.model_validate(budget)


@router.patch("/budgets/{budget_id}", response_model=BudgetResponse)
async def update_budget(
    budget_id: uuid.UUID,
    data: BudgetUpdate,
    ctx: VisibilityContext = Depends(get_visibility_ctx),
    session: AsyncSession = Depends(get_session),
) -> BudgetResponse:
    svc = BudgetService(session)
    budget = await svc.update(ctx, budget_id, data)
    await session.commit()
    await session.refresh(budget)
    return BudgetResponse.model_validate(budget)


@router.delete("/budgets/{budget_id}", status_code=204)
async def delete_budget(
    budget_id: uuid.UUID,
    ctx: VisibilityContext = Depends(get_visibility_ctx),
    session: AsyncSession = Depends(get_session),
) -> None:
    svc = BudgetService(session)
    await svc.delete(ctx, budget_id)
    await session.commit()
